### PR TITLE
GTSAM_DT_MERGING Flag

### PR DESCRIPTION
**Pull request by varunagrawal**
~~This PR is a discussion venue for issue #1496.~~

~~I have added a unit test which demonstrates an issue with the number of assignments being set for the decision tree.
This seems to be because the decision tree rearranges the keys to be highest to lowest but doesn't account for the implicit pruning performed. If I disable the implicit pruning in the decision tree, this issue doesn't appear.~~

This PR adds a new CMake flag `GTSAM_DT_MERGING` which enables/disables `DecisionTree` merging of branches when the leaves have the same value. This flag is generally useful for debugging and maybe useful for other things.

Accordingly I have updated the tests and the config files.

I have also removed the `nrAssignments_` variable in `DecisionTree` due to it needing significant refactor and the issue it was originally intended for (hybrid pruning) being resolved via alternative means.

### CODE REVIEW

File: gtsam/discrete/DecisionTree.h [hunk 1]
```diff
@@ -299,6 +299,15 @@ namespace gtsam {
     /// Return the number of leaves in the tree.
     size_t nrLeaves() const;
 
+    /**
+     * @brief Return the number of total leaf assignments.
```
Comment (dellaert) [inline gtsam/discrete/DecisionTree.h:303]: Please give examples as to what this is, I think "number of assignments" needs to be explained.

**Comment by dellaert** [comment]
@varunagrawal I think we discussed this and is no more relevant? If so, please close issue and PR. If not, let's chat again as I lost context.

**Comment by varunagrawal** [comment]
This is still a bug, one that I introduced accidentally.

**Comment by varunagrawal** [comment]
So I managed to figure this out after a *lot* of debugging.

The big issue was that when building the tree, we call `Unique` which does the pruning, but once the tree is built, pruned nodes are added back if the labels are not in decreasing order (this is handled by `compose`).

To address this, I made `Unique` a recursive method that now operates only after the whole tree is built and the labels reordered. The tree construction process is now: build tree bottom-up, order labels, prune top-down. The previous approach was: build *and* prune bottom-up, order labels.

**Comment by varunagrawal** [comment]
#1555 details some of the pending issues I have with `nrAssignments`. I'll get back to those after I update Hybrid to use the TableFactor.

**Comment by dellaert** [comment]
Please comment on what the status is of this PR after you merged all other 3 we discussed. Also, might want to merge in develop after that to make this PR much smaller?

**Comment by varunagrawal** [comment]
Merged in develop. The `nrAssignments` is still messed up due to the binary `apply` method, and I'll push a unit test to this PR so we can track it later.

**Comment by varunagrawal** [comment]
This PR is still a WIP. I'll update it as I go along.

**Comment by varunagrawal** [comment]
I removed `nrAssignments` since the issues go much deeper and doing a full fix would take far too long, time which I currently don't have. Also the original reason I added that feature was to improve the pruning speed, but that was accomplished in another way.

**Review by dellaert — APPROVED** [review]
OK, I like the simplification.

**Pull request merged**